src/dns/query.cpp

```
@@ -3,46 +3,306 @@
+        answer.ttl = ttl;
+        answer.type = QueryTypeId::PTR;
+        // Note: cast magic copied from libevent regress tests
+        answer.hostname = std::string(*(char **)addresses);
+        logger->info("dns - adding %s", answer.hostname.c_str());
+        answers.push_back(answer);
+    } else if (type == DNS_IPv4_A || type == DNS_IPv6_AAAA) {
+
+        int family;
+        int size;
+        char string[128]; // Is wide enough (max. IPv6 length is 45 chars)
+
+        if (type == DNS_IPv4_A) {
+            family = PF_INET;
+            size = 4;
+            logger->info("dns - IPv4");
+        } else {
+            family = PF_INET6;
+            size = 16;
+            logger->info("dns - IPv6");
+        }
+
+        //
+        // Note: make sure in advance `i * size` won't overflow,
+        // this is here only for robustness.
+        //
+        if (count >= 0 && count <= INT_MAX / size + 1) {
+
+            for (auto i = 0; i < count; ++i) {
+                // Note: address already in network byte order
+                if (inet_ntop(family, (char *)addresses + i * size,
+                              string, sizeof(string)) == NULL) {
+                    logger->warn("dns - unexpected inet_ntop failure");
+                    code = DNS_ERR_UNKNOWN;
+                    break;
+                }
+                Answer answer;
+                answer.code = code;
+                answer.ttl = ttl;
+                if (family == PF_INET) {
+                    answer.ipv4 = string;
+                    answer.type = QueryTypeId::A;
+                } else if (family == PF_INET6) {
+                    answer.ipv6 = string;
+                    answer.type = QueryTypeId::AAAA;
+                }
+                logger->info("dns - adding '%s'", string);
+                answers.push_back(answer);
+            }
+
+        } else {
+            logger->warn("dns - too many addresses");
+            code = DNS_ERR_UNKNOWN;
```

Also in this case I'd throw an exception because again this is a corner case.